### PR TITLE
C++20 co_await support for Embind promises

### LINKED ISSUE #20413: C++20 coroutines + Native webassembly promise integration
From side of C++20 we have coroutines, from side of WebAssembly (at least in Chrome) we have native promise integration. Why be not to unify such features, and not to make an new and better async await API? Probably, make some sort of wrap or emulation? Or coroutine header with low-level code implementation?

### PR DESCRIPTION
**Pull request by RReverser**
This adds support for `co_await`-ing Promises represented by `emscripten::val`.

The surrounding coroutine should also return `emscripten::val`, which will be a promise representing the whole coroutine's return value.

Note that this feature uses LLVM coroutines and so, doesn't depend on either Asyncify or JSPI. It doesn't pause the entire program, but only the coroutine itself, so it serves somewhat different usecases even though all those features operate on promises.

Nevertheless, if you are not implementing a syscall that must behave as-if it was synchronous, but instead simply want to await on some async operations and return a new promise to the user, this feature will be much more efficient.

Here's a simple benchmark measuring runtime overhead from awaiting on a no-op Promise repeatedly in a deep call stack:

```cpp

using namespace emscripten;

// clang-format off
EM_JS(EM_VAL, wait_impl, (), {
  return Emval.toHandle(Promise.resolve());
});
// clang-format on

val wait() { return val::take_ownership(wait_impl()); }

val coro_co_await(int depth) {
  co_await wait();
  if (depth > 0) {
    co_await coro_co_await(depth - 1);
  }
  co_return val();
}

val asyncify_val_await(int depth) {
  wait().await();
  if (depth > 0) {
    asyncify_val_await(depth - 1);
  }
  return val();
}

EMSCRIPTEN_BINDINGS(bench) {
  function("coro_co_await", coro_co_await);
  function("asyncify_val_await", asyncify_val_await, async());
}
```

And the JS runner also comparing with pure-JS implementation:

```js
import Benchmark from 'benchmark';
import initModule from './async-bench.mjs';

let Module = await initModule();
let suite = new Benchmark.Suite();

function addAsyncBench(name, func) {
	suite.add(name, {
		defer: true,
		fn: (deferred) => func(1000).then(() => deferred.resolve()),
	});
}

for (const name of ['coro_co_await', 'asyncify_val_await']) {
  addAsyncBench(name, Module[name]);
}

addAsyncBench('pure_js', async function pure_js(depth) {
  await Promise.resolve();
  if (depth > 0) {
    await pure_js(depth - 1);
  }
});

suite
  .on('cycle', function (event) {
    console.log(String(event.target));
  })
  .run({async: true});
```

Results with regular Asyncify (I had to bump up `ASYNCIFY_STACK_SIZE` to accomodate said deep stack):

```bash
> ./emcc async-bench.cpp -std=c++20 -O3 -o async-bench.mjs --bind -s ASYNCIFY -s ASYNCIFY_STACK_SIZE=1000000
> node --no-liftoff --no-wasm-tier-up --no-wasm-lazy-compilation --no-sparkplug async-bench-runner.mjs

coro_co_await x 727 ops/sec ±10.59% (47 runs sampled)
asyncify_val_await x 58.05 ops/sec ±6.91% (53 runs sampled)
pure_js x 3,022 ops/sec ±8.06% (52 runs sampled)
```

Results with JSPI (I had to disable `DYNAMIC_EXECUTION` because I was getting `RuntimeError: table index is out of bounds` in random places depending on optimisation mode - JSPI miscompilation?):

```bash
> ./emcc async-bench.cpp -std=c++20 -O3 -o async-bench.mjs --bind -s ASYNCIFY=2 -s DYNAMIC_EXECUTION=0
> node --no-liftoff --no-wasm-tier-up --no-wasm-lazy-compilation --no-sparkplug --experimental-wasm-stack-switching async-bench-runner.mjs

coro_co_await x 955 ops/sec ±9.25% (62 runs sampled)
asyncify_val_await x 924 ops/sec ±8.27% (62 runs sampled)
pure_js x 3,258 ops/sec ±8.98% (53 runs sampled)
```

So the performance is much faster than regular Asyncify, and on par with JSPI.

Fixes #20413.

### CODE REVIEW

**Comment by RReverser** [comment]
TODO for myself - add docs section and Changelog; just wanted to start upstreaming this sooner rather than later, since I've been playing with it for some time and looks like other people want it too.

**Comment by unit-404** [comment]
Why `val`? Why not to make from promise argument or returned `em_promise_t` (as version)? I shown version about? Also, explain how `void** result` should to working.

**Comment by RReverser** [comment]
`em_promise_t` is a separate type with its own promise management API, its own handle allocator etc, whereas Embind is de-facto standard API for operating on arbitrary JavaScript values.

I needed to await JavaScript promises retrieved from Embind, so that's what I implemented, but if you want to implement coroutines for `em_promise_t`, you can also do so separately.

**Comment by RReverser** [comment]
> Also, explain how `void** result` should to working.

But also... I don't know what you mean by this. Do you mean receiving arbitrary/untyped values from JS? If so, `val` is what you want.

**Comment by RReverser** [comment]
Not sure where this is coming from in wasm64:

```
TypeError: Cannot convert 8 to a BigInt
    at test_val_coro.wasm.asyncCoro() (wasm://wasm/test_val_coro.wasm-000192de:wasm-function[26]:0x6df)
    at test_val_coro.wasm.emscripten::internal::Invoker<emscripten::val>::invoke(emscripten::val (*)()) (wasm://wasm/test_val_coro.wasm-000192de:wasm-function[34]:0xbdd)
    at dynCall (C:\Users\me\Documents\emscripten\out\test\test_val_coro.js:1414:40)
    at C:\Users\me\Documents\emscripten\out\test\test_val_coro.js:1422:16
```

Given the low number, I suspect it's one of the value handles that it tries to pass to JS where `i64` is expected, but it's unclear to me why is it a 32-bit number in the first place if `EM_VAL` are just pointers.

**Comment by RReverser** [comment]
Seems to be failing on attempt to call `promise_sleep_impl` in test itself:

image

Its declaration looks ok though:

image

Maybe `Emval.toHandle` (which returns integer) is not correctly converted in wasm64 mode? If so, the implications / failures would be wider than just my test...

**Comment by RReverser** [comment]
Ah yeah, that's probably it - it appears no other tests are doing `EM_JS(EM_VAL, ...)` so it was simply not caught yet.

**Comment by RReverser** [comment]
Yup, can reproduce with a separate test. Going to submit a fix.

**Comment by RReverser** [comment]
See https://github.com/emscripten-core/emscripten/pull/20421.

**Comment by sbc100** [comment]
@tlively who is the maintainer of the current C/C++ promise integration.

**Comment by RReverser** [comment]
I'll add that one particular motivation for this is that it's impossible to use proxying APIs with Asyncify at the moment; even if it was, it would be difficult because different threads could be competing for Asyncify state which can't handle queueing.

So when I needed to do something like

```cpp
proxySync([] {
  val foo = doSomethingSynchronous();
  val bar = foo.call<void>("someAsyncMethod").await();
  val baz = somethingSynchronousAgain(bar);
  ...another await..
});
```

it had to be rewritten so that each synchronous call is proxied separately in one `proxySync`-based helper, and each asynchronous call has to be wrapped into another helper that subscribes to promise and uses `proxySyncWithCtx`, so it kept going forth and back between threads in a rather ugly mix of code.

With coroutines this is not a problem, as I can proxy the entire coroutine execution to the main thread in one go and use `co_await` for awaiting those intermediate values.

**Comment by RReverser** [comment]
Ping, would love a review on this. I have a project where having `co_await` would come very handy.

File: site/source/docs/api_reference/val.h.rst
Comment (brendandahl) [inline site/source/docs/api_reference/val.h.rst:267]: Change to: The ``co_await`` operator allows awaiting on JavaScript promises represented by ``val``.

File: site/source/docs/api_reference/val.h.rst
Comment (brendandahl) [inline site/source/docs/api_reference/val.h.rst:272]: Change to: For example, it allows you to implement the equivalent of this JavaScript...

File: system/include/emscripten/val.h
Comment (brendandahl) [inline system/include/emscripten/val.h:116]: Should we build siginfo with c++20?

File: test/embind/test_val_coro.cpp
Comment (brendandahl) [inline test/embind/test_val_coro.cpp:26]: s/receving/receiving

**Review by brendandahl — COMMENTED** [review]
Having not used c++20 coroutines at all yet, this seems reasonable and a pretty nice feature to avoid asyncify/jspi. I'd still like to hear @tlively's thoughts since he had started doing something with coroutines.

File: site/source/docs/api_reference/val.h.rst
Comment (RReverser) [reply site/source/docs/api_reference/val.h.rst:267]: Small ask if you don't mind: could you please use Github's "Add a suggestion" block for small suggestions like these? Then I could quickly accept them directly from UI.

File: system/include/emscripten/val.h
Comment (RReverser) [reply system/include/emscripten/val.h:116]: Ah would that help / can we do that? If so, maybe! I never looked into how these run.

File: site/source/docs/api_reference/val.h.rst
Comment (RReverser) [reply site/source/docs/api_reference/val.h.rst:267]: I'll do it myself for these few.

```suggestion
   The ``co_await`` operator allows awaiting JavaScript promises represented by ``val``.
```

File: site/source/docs/api_reference/val.h.rst
Comment (RReverser) [reply site/source/docs/api_reference/val.h.rst:272]:
```suggestion
   For example, it allows you to implement the equivalent of this JavaScript ``async``/``await`` function:
```

File: test/embind/test_val_coro.cpp
Comment (RReverser) [reply test/embind/test_val_coro.cpp:26]:
```suggestion
  // check that sleeping and receiving value works
```

File: system/include/emscripten/val.h
Comment (RReverser) [reply system/include/emscripten/val.h:116]: Seems pretty straightforward... I'll try that now.

**Comment by RReverser** [comment]
> and a pretty nice feature to avoid asyncify/jspi

@brendandahl Btw, just wanted to add that I found it has its place in combination with Asyncify too, not necessarily instead of.

In particular, it helps to write a helper coroutine that does a bunch of `co_await` that's handled by C++ compiler, and then do a single `my_func().await()` to wait for all of them to finish, this time with Asyncify.

This way, when you need to pause the entire app, you can do so with just one expensive unwind/rewind by Asyncify instead of having lots of individual unwinds/rewinds for every async operation inside that function.

**Comment by sbc100** [comment]
I'm afraid I haven't yet understood C++ co-routines or how that can be useful here.   I would like to take some time to better understand this change, but if @tlively or @brendandahl think they have a good handle on it I'll defer to them.

**Comment by tlively** [comment]
Meanwhile I don't understand any of the embind stuff, nor do I understand the C++20 co_await stuff off the top of my head 😅 

I'll have to dig into cppreference to understand the coroutine stuff here. @brendandahl and @sbc100, do you want to schedule some time to review this collaboratively?

**Comment by RReverser** [comment]
Lol I love where this is going 😅 

To be fair, I didn't know anything about C++ coroutines / co_await before trying to implement it either, and the terminology C++ used for this stuff is confusing / different from other languages, which didn't help... but in the end it works.

> @brendandahl and @sbc100, do you want to schedule some time to review this collaboratively?

FWIW I'd be happy to join in for a meeting if that helps, assuming Europe-friendly time.

**Comment by sbc100** [comment]
> Lol I love where this is going 😅
> 
> To be fair, I didn't know anything about C++ coroutines / co_await before trying to implement it either, and the terminology C++ used for this stuff is confusing / different from other languages, which didn't help... but in the end it works.
> 
> > @brendandahl and @sbc100, do you want to schedule some time to review this collaboratively?
> 
> FWIW I'd be happy to join in for a meeting if that helps, assuming Europe-friendly time.

Sounds good.  How about Friday morning.  9am PST?

**Comment by RReverser** [comment]
> How about Friday morning. 9am PST?

Hm so it will be Friday evening here... tentatively yes, but if we can do that in first half of the week, would be a bit better.

**Comment by RReverser** [comment]
> Sounds good. How about Friday morning. 9am PST?

Should we give it a try this week to help unblock this?

**Comment by sbc100** [comment]
> > Sounds good. How about Friday morning. 9am PST?
> 
> Should we give it a try this week to help unblock this?

9am PST tomorrow?  sgtm.. I'll send out an calendar invite.

File: system/include/emscripten/val.h
Comment (tlively) [inline system/include/emscripten/val.h:695]: Can we move all this to the top of the class since understanding this state is necessary for understanding the rest of the class?

File: test/embind/test_val_coro.cpp
Comment (tlively) [inline test/embind/test_val_coro.cpp:34]: Can we make this message less generic? When I read the test Python, this looked so generic that I thought it was coming from libc++ somehow. Maybe `"bang!"` or similar.

File: src/library_sigs.js
Comment (tlively) [inline src/library_sigs.js:343]: Wouldn't it be nicer to put the `__sig` alongside the source? Why do we have this separate sigs file?

**Review by tlively — APPROVED** [review]

File: src/library_sigs.js
Comment (RReverser) [reply src/library_sigs.js:343]: It's autogenerated by tooling from C/C++.

File: system/include/emscripten/val.h
Comment (RReverser) [reply system/include/emscripten/val.h:706]: Not really, I mean they're almost the same except for privacy defaults . I'll change.

File: system/include/emscripten/val.h
Comment (RReverser) [reply system/include/emscripten/val.h:695]: Yeah TBH I don't understand why the common style is to put private items in the end of the class.

**Comment by RReverser** [comment]
Decided to add a bunch of comments explaining what those special coroutine types & methods do. They somewhat duplicate generic docs about C++ coroutines, but figured they might be useful in code as probably few people have to ever implement them.

File: system/include/emscripten/val.h
Comment (RReverser) [reply system/include/emscripten/val.h:677]: AFAIK yup :(

Comment (RReverser) [reply system/include/emscripten/val.h:677]: I mean, if you don't have duplicate types, then you can at least avoid explicit indices as it will "compute" them for you, but that wouldn't work here as 2 states both use `val`.

Comment (sbc100) [reply system/include/emscripten/val.h:677]: Its strange they don't have version that is basically as safer/checked/tagged version of the C "union" where each member can have a name and the syntax is simple to read like the "C" union.

Comment (sbc100) [reply system/include/emscripten/val.h:677]: Resolving this..

Comment (RReverser) [reply system/include/emscripten/val.h:677]: > Its strange they don't have version that is basically as safer/checked/tagged version of the C "union" where each member can have a name and the syntax is simple to read like the "C" union.

Tell me about it... I guess they're hesitant to add more syntax and try to implement everything via STL, but the result is unpleasant.

**Comment by RReverser** [comment]
Can someone push new docs to the website please? @kripken IIRC you had to do that manually in the past - is that still the case?

**Comment by kripken** [comment]
Yes. Updated now!

**Comment by andreamancuso** [comment]
Hi, this looks great, I'm getting my hands dirty with coroutines and was wondering if you were planning to add coroutine support for `-sFETCH` ?

I have been trying to parallelize HTTP requests and do `when_all()` - I enabled `pthreads` but am still getting 
image

**Pull request merged**
